[2022] Prepoznavanje govora (#4)
pfe-rs/izvestaji: content/2022/prepoznavanje-govora.md

title: Prepoznavanje govora
summary: Projekat iz prepoznavanja govora rađen na letnjem kampu za stare polaznike 2022. godine od Dimitrija Pešića i Lazara Zubovića.
---
### Apstrakt

### Apstrakt na engleskom
### Uvod

Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. Motivacija projekta bila je u tome da se ne samo primene mnoge metode korišćene za speech recognition, već da se i uporede njihova praktičnost i upotreba u praktičnim svrhama. 

> Projekat se zasniva na ideji korišćenja spektrograma kao osnovne metode prikaza zvuka u 2D formatu. Iz tog formata, drugačijim metodama bi se zvuk prepoznavao sa spektrograma što je zapravo ništa drugo no obična slika. Sa te slike mogu se pokupiti različiti podaci o zvuku zarad preciznijeg prepoznavanja istog.

Reviewer: Tacno je da uvek vase metode koriste spektrogram, ali je malko missleading da pricate u celom pasosu o spektrogramu kao o slici. Kod MFCC metode, ne gledate spektrogram kao sliku, vec iz njega rucno izvlacite set feature-a. Razdvojite jasno te pristupe

> Prelazeći kroz literaturu i referentne radove, mnogi su više doprineli pri metodi obrade spektrograma nego pri izradi samih spektrograma. 

Reviewer: ```suggestion
Prelazeći kroz literaturu i referentne radove, mnogi su više doprineli pri metodi obrade spektrograma nego pri izradi samih spektrograma. 
```
Obrisao bih ovu recenicu

> Osvrt na rad ogleda se u metodama koje su pokrivene u referentnim radovima, poput MFCC-a (Mel-Frequency Cepstral Coefficients), logističke regresije, Random Forest-a, SVM-a, XGBoost-a, kao i konvolucionih neuronskih mreža. Do ovog projekta, ljudi su fokusirali svoje radove na obradi jednog metoda i testiranju na određenoj bazi. Nasuprot njihovim, ovaj rad ima dosta limitiranu bazu, te i sami rezultati variraju u odnosu na već dobijene.

Reviewer: ```suggestion
Ajde da razdvojimo dve celine. Postoji set metoda kod kojih imamo deo koji se bavi ekstrakcijom ficera (to je MFCC), pa potom primenimo klasifikator (kao xgboost, logisticka reg, ...). U drugi set metoda spadaju konvolucione neuronske mreze, kod kojih se izvlacenje ficera ne vrsi rucno, vec mreza sama nauci koje ficere da izvuce, a takodje vrsi i klasifikaciju. Nemojte da govorite o metodama i nabrojite svaki strucni pojam iz projekta. Bitno je razdvojiti sta je metoda za izvlacenje ficera, sta klasifikator, a sta oba 
```

> Osvrt na rad ogleda se u metodama koje su pokrivene u referentnim radovima, poput MFCC-a (Mel-Frequency Cepstral Coefficients), logističke regresije, Random Forest-a, SVM-a, XGBoost-a, kao i konvolucionih neuronskih mreža. Do ovog projekta, ljudi su fokusirali svoje radove na obradi jednog metoda i testiranju na određenoj bazi. Nasuprot njihovim, ovaj rad ima dosta limitiranu bazu, te i sami rezultati variraju u odnosu na već dobijene.

Reviewer: ```suggestion
Ajde da razdvojimo dve celine. Postoji set metoda kod kojih imamo deo koji se bavi ekstrakcijom ficera (to je MFCC), pa potom primenimo klasifikator (kao xgboost, logisticka reg, ...). U drugi set metoda spadaju konvolucione neuronske mreze, kod kojih se izvlacenje ficera ne vrsi rucno, vec mreza sama nauci koje ficere da izvuce, a takodje vrsi i klasifikaciju. Nemojte da govorite o metodama i nabrojite svaki strucni pojam iz projekta. Bitno je razdvojiti sta je metoda za izvlacenje ficera, sta klasifikator, a sta oba.
Takodje, ova recenica nije tacna "Do ovog projekta, ljudi su fokusirali svoje radove na...".
```

> Osvrt na rad ogleda se u metodama koje su pokrivene u referentnim radovima, poput MFCC-a (Mel-Frequency Cepstral Coefficients), logističke regresije, Random Forest-a, SVM-a, XGBoost-a, kao i konvolucionih neuronskih mreža. Do ovog projekta, ljudi su fokusirali svoje radove na obradi jednog metoda i testiranju na određenoj bazi. Nasuprot njihovim, ovaj rad ima dosta limitiranu bazu, te i sami rezultati variraju u odnosu na već dobijene.

Reviewer: Takodje, niste odradili pregled literature. To podrazumeva da pogledate i citirate druge naucne radove i kako su se oni bavili ovakvom problematikom. Naravno, mislim na klasifikaciju reci, jer je to sustinski vas projekat, ne speech recognition u pravom smislu reci

> ### Aparatura i metoda

Reviewer 2: Mislim da je nepotrebno da ovaj naslov sadrži i reč "aparatura", s obzirom da je vi, u suštini, nemate (osim ako ne planirate da posvetite ovaj deo tome da navodite aparaturu korišćenu za deo projekta u kom ste vi snimali podatke za svoju bazu podataka). Takođe, bitno nam je da imamo uvid u to koji su podaci korišćeni, pa je skroz okej pre celog ovog dela dodati i segment "Podaci" ili nešto slično gde ćete govoriti o obe baze podataka koje ste koristili: nad čime ste vi zapravo radili prepoznavanje govora, koliko podataka imamo, ...

Rešenje datog problema prepoznavanja govora svodi se na izradu spektrograma i obradu istih.

#### Spektrogrami

> Spektrogrami su vizuelne reprezentacije jačine signala, to jest glasnoće zvuka u nekom vremenskom intervalu. Mogu se posmatrati kao dvodimenzionalni grafici gde se može uočiti i treća dimenzija preko boja svakog dela spektrograma. Vremenska osa se gleda sa leve na desnu stranu po horizontalnoj osi. Vertikalna osa predstavlja frekvenciju koju možemo posmatrati i kao ton zvuka. U logaritamskoj je skali kako bi se prilagodila ljudskom uhu koje čuje po istom principu.

Reviewer: Okej, izbacicemo izraz "glasnoća". Objasnite sta znaci ta frekvencijska osa. Spektrogram prikazuje kolika je amplituda svake frekvencijske komponente signala, za svaki vremenski interval. Ti intervali su mali, jer na malim intervalima mozemo da pretpostavimo da se amplitude frekvencijskih komponenti ne menjaju. Ovaj deo posle je okej, recite samo jasno da boja oznacava amplitudu signala

> Boja na grafiku predstavlja amplitudu zvuka u određenom vremenskom trenutku. Plava boja na spektrogramu predstavlja niske amplitude, dok crvena boja predstavlja visoke amplitude.

Reviewer: Evo ovde ste to lepo rekli

Primena spektrograma u ovom radu jeste prepoznavanje fonema reči kako bi, spajanjem istih, reč mogla da se prepozna.

> U Python programskom jeziku, zvuk se može transformisati u spektrogram korišćenjem biblioteke Librosa. Ova biblioteka se koristi pri rešavanju problema sa analizom fajlova audio formata. 

Reviewer: Ne morate bas konkretne biblioteke da referencirate (pogotovo kad ne pricamo o nekim super poznatim kao numpy pytorch sklearn i sl.). Dovoljno je da napisete da je projekat realizovan u pythonu

#### Metode obrade spektrograma

##### 1. Logistička regresija

> Logistička regresija je metoda klasifikacije koja se može primeniti i koristiti svuda gde imamo promenljive koje se mogu kategorisati. Za razliku on linearne regresije, vrednosti njenih rezultata su ograničene između 0 i 1. 

Reviewer: typo drugi red pred kraj

> Ova metoda za klasifikaciju ne koristi linearnu već sigmoidnu funkciju bilo kog tipa, a primer takve funckije je dat na slici 1.

Reviewer: Koristi sigmoidnu u slucaju binarne klasifikacije, odnosno softmax u slucaju vise klasa

![Sigmoid](static\images\1.png)

> Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. To često nije dovoljno, pa se koristi multinomijalna logistička regresija (ili Softmax Regression) koja može da razlikuje više od dve različite kategorija.

Reviewer: Typo poslednji red poslednja rec

> Funkcija cene ove metode je logaritamska kako bi se dobila konveksna završna funkcija parametara i time se postiglo da gradient descent nađe globalni, a ne samo lokalni minimum funkcije.

Reviewer 2: Kriterijumska funkcija je bolji termin od funkcije cene. 

Ne razumem baš šta podrazumevate pod završnom funkcijom parametara. Mi želimo da minimizujemo funkciju koja opisuje grešku našeg modela odnosno kriterijumsku funkciju, pa je samim tim ona ta koja je konveksna u slučaju multinomijalne logističke regresije.

![Funkcija](static\images\2.png)

- hΘ(x) = sigmoid (w*x + b), Y rezultat, x = promenljiva koju posmatramo

> Da bi logistička regresija dala što bolje rezultate, trenira se MLE (Maximum Likelihood Estimation) metodom, nameštajući beta parametre kroz više iteracija tražeći najbolje fitovanu krivu, odakle se biraju najbolje procene parametara. Nakon toga se dobijeni koeficijenti koriste za računanje verovatnoće za svaki primer, pa se one logaritmuju i sabiraju i time formiraju konačnu predviđenu verovatnoću. Svaka vrednost iznad 0.5 (ili bilo koje zadate granice) se tretira kao da je jedinica, a svaka manja od te granice se tretira kao nula.

Reviewer 2: Šta su beta parametri? 

Objasnili ste kako funkcioniše logistička regresija kada imamo slučaj 2 kategorije. Međutim, kako ste naveli da vi u vašem radu koristite softmax regresiju, trebalo bi da govorite o verovatnoćama u tom slučaju. Ako imamo K kategorija (u vašem slučaju 10), mi želimo da napravimo softmax regresiju koja će nam zapravo dati K različitih verovatnoća, čiji je zbir 1. Potom, konačnu odluku donosimo po tome koja kategorija nosi najveću verovatnoću za zadate ulazne podatke.

##### 2. MFCCs

MFCCs (Mel-Frequency Cepstral Coefficients) jesu koeficijenti koji opisuju karakteristike na osnovu spektrograma određenog zvuka. Njihova primena u ovom projketu svodi se na izdvajanje ključnih odlika nekog zvuka kako bi reč mogla da se prepozna. Te odlike se zovu formonti i njih stvara ljudski vokalni trakt prilikom govora, menjajući čist glas koji stvaraju naše glasne žice dok vibriraju. Ove odlike se formiraju u reč.

Kepstar (cepstrum) je spektar spektra. On nastaje inverznom Furijeovom transformacijom logaritmovanog spektra. Formula za nastanak kepstra:

$C(x(t))=F^{-1}[\log (F[x(t)])]$

Proces stvaranja kepstra je sledeći:

1. Na dobijeni signal primenimo diskretnu Furijeovu transformaciju. Ova transformacija nam daje grafik zavisnosti jačine zvuka od frekvencije po sledećoj formuli:

$\begin{aligned} X_k &=\sum_{n=0}^{N-1} x_n \cdot e^{-\frac{i 2 \pi}{N} k n} \\ &=\sum_{n=0}^{N-1} x_n \cdot\left[\cos \left(\frac{2 \pi}{N} k n\right)-i \cdot \sin \left(\frac{2 \pi}{N} k n\right)\right] \end{aligned}$

> 2. Power spektar logaritmujemo, pa odatle dobijamo spektar koji na vertikalnoj osi pokazuje jačinu zvuka u decibelima (dB), a horizontalna osa i dalje prikazuje frekvenciju.

Reviewer: Ajmo na srpskom - spektar snage. Objasnite sta je spektar snage

![Power spectar](static\images\log.png)

3. Po logaritmovanju power spektra, izvršenjem inverzne Furijeove transformacije dobijamo kepstar.

> Prednost kepstra i Mel-Frequency kepstra jeste u sličnosti y-ose sa ljudskim glasom. Ljudski glas se odlikuje u jačini zvuka koja je logaritamska veličina, kao i kod kepstara.

Reviewer: Objasnite sta je mel frequency skala (i probajte da posrbite taj izraz). Objasnite da je linearna do otp 1kHz i da posle ima logaritamsku zavisnost.

Mel filter banke ...

> U Pythonu, implementacija MFCC-a svodi se na lični odabir koliko odlika zvuka je potrebno izvući za precizna predviđanja. Librosa biblioteka dalje obogućava obradu zvuka kroz kepstre i izvlačenje traženih odlika.

Reviewer: Nije relevantno, brisite slobodno

##### 3. Random Forest

> Random Forest je klasifikator koji koristi više stabala odlučivanja (Desicion Tree) i njihova pojedinačna predviđanja stapa u jedno konačno.

Reviewer: Desicion -> Decision

Za treniranje mreže koriste se dve metode simultano: loss funkcija i back propagation.

Cost funkcija (funkcija troškova ili gubitka) jeste funkcionalna veza željenog outputa i dobijenog outputa u funkciji, a loss funkcija je srednja vrednost svih cost funkcija.

Najkorišćenija loss funkcija je Cross Entropy Loss. 
Cross Entropy Loss radi tako što pokušava da minimizuje razliku između tačnih rezultata i verovatnoće predviđanja, to jest output.

Formula po kojoj se računa Cross Entropy Loss je sledeća:

$H_p(q)=-\frac{1}{N} \sum_{i=1}^N y_i \cdot \log \left(p\left(y_i\right)\right)+\left(1-y_i\right) \cdot \log \left(1-p\left(y_i\right)\right)$

Back propagacija je metod smanjenja grešaka u CNN posmatranjem neophodnih promena u prethodnom sloju od aktivacije da bi se u određenom sloju neuroni aktivirali na određen način.

Backpropagation prolazi krroz sve primere i traži sumu svih težina veza među neuronima.

Težine se menjaju u cilju računanja dovoljno dobrog gradient descenta za traženje lokalnog / maksimalnog minimuma ove funkcije, to jest tačnu reč.

![SGD](static\images\sgd.png)

> ### Istraživanje i rezultati

Reviewer 2: Navedite koliko ste podataka koristili za treniranje i testiranje i eventualno za validaciju pri treniranju CNN-om.

> Rezultati su krajnje očekivani uzimajući u obzir veličinu baze koja je obrađivanja. Bez interaktivnog interfejsa, dosadašnji rezultati svode se na tačnost (accuracy) svake metode u radu. 

Reviewer: Ovaj pasos mi je ceo nepotreban. Aj ovako: U tabeli toj i toj mozemo videti rezultate. Kao metrika izbora koriscena je tacnost, s obzirom da je baza podataka izbalansirana, sto cini tacnost reprezentativnom metrikom. Kao sto se moze videti, CNN metoda postize znacajno bolje rezultate od drugih metoda. Medju metodama koje ne koriste duboko ucenje, XGBoost daje najbolje rezultate.
Eeee sad kad ste jedan ceo pasos objasnili sta pise u tabeli, mozete da krenete u diskusiju da li se rezultati slazu ili ne sa hipotezama. Izbegavajte izraz "ocekivano je", vec kao sto je istrazeno tu i tu, ...

![Rezultati](static\images\4.png)

> Iz tabele iznad može se uočiti kako rezultati dosta variraju jedni od drugih. Konvoluciona neuronska mreža daje maksimalnu preciznost u istim uslovima, dok SVM sa polinomijalnim kernelom daje minimalne. 

Reviewer 2: Preciznost je metrika koja se definiše na drugačiji način u odnosu na tačnost tako da pazite da ne mešate ta dva termina.

> Svoj potencijal SVM može da pokaže kada je lako odrediti kojoj labeli koji podatak pripada. U ovom slučaju, određene reči mogu lako da se pomešaju na spektrogramu, pa su neke vrednosti vrlo blizu odlučnoj granici i da pomute labele. Iz tog razloga, rezultati su veoma dobri za ovu metodu. 

Reviewer: Zar ne rade svi klasifikatori najbolje kad je lako odrediti labelu podatka hahahahah?
Btw ne bih se upustao u analizu zasto je linearni kernel dao bolje rezultate od polinomskog. U teoriji to stvarno ne bi smelo da se desi, tkd je vrv neka glupa implementacijska greska

SVM daje različite rezultate u zavisnosti od svojih kernela. Linearni kernel se najbolje pokazao zato što se usaglašava sa zadatkom koji mu je dat (svaka reč ima dosta odlika na osnovu kojih se labelira), a i u ovom kernelu potrebno je samo da optimizujemo C Regularisation parametar, pa je treniranje brže.

> Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Deep learning metode povoljnije su za feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao speech recognition loss funkcija pošto ljudsko uho reaguje logaritamski. Gledajući ova dva faktora u obzir, očekivano je da će performansa CNN-a biti najbolja.

Reviewer 2: Šta znači ovo: "ljudsko uho reaguje logaritamski"?

> Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Deep learning metode povoljnije su za feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao speech recognition loss funkcija pošto ljudsko uho reaguje logaritamski. Gledajući ova dva faktora u obzir, očekivano je da će performansa CNN-a biti najbolja.

Reviewer: 1. Pre bih rekao da se kod metoda dubokog ucenja proces ekstrakcije karakteristika (da posrbimo) vrsi u okviru same metode, te nije potrebno rucno biranje karakteristika.

Razlika između regresora i klasifikatora objašnjavaju se samom ulogom regresora i klasifikatora pri povezivanju određenih podataka sa njihovim labelama.

> Regresori imaju veću primenu kada je potrebno neku tačnu vrednost dati kao labelu nekom podatku, dok klasifikator stavlja podatak u određenu kategoriju i tako daje labelu. U slučaju speech recognitiona, u FSDD bazi dato je 10 labela, pa klasifikator radi bolji posao da pretpostavi u koju kategoriju labela određeni zvuk spada (cifra od 0 do 9).

Reviewer 2: Nije mi baš jasno zašto ste dodali ovaj i prethodni pasus gde ste pravili razliku između regresije i klasifikacije. Mislim da bi trebalo da bude očigledno još u uvodu da vi radite na problemu kategorizacije reči pa je stoga ovo što ste napisali redundantno u kontekstu "Istraživanja i rezultata".

title: Prepoznavanje govora
summary: Projekat iz prepoznavanja govora rađen na letnjem kampu za stare polaznike 2022. godine od Dimitrija Pešića i Lazara Zubovića.
---
### Apstrakt

### Apstrakt na engleskom
### Uvod

> Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. Prepoznavanje govora je proces osposobljavanja nekog modela da prepozna i odreaguje na zvuk proizveden ljudskim govorom. Model uzima audio signal u formi talasa, izvlači iz njega podatke, obrađuje ih, prepoznaje i preduzima određene korake u zavisnosti od rezultata.

Reviewer 2: ```suggestion
Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. Prepoznavanje govora je proces osposobljavanja nekog modela da prepozna i odreaguje na zvuk proizveden ljudskim govorom. Model uzima audio signal u formi talasa, izvlači iz njega podatke, obrađuje ih i identifikuje izgovorenu reč.
```

Reviewer 2: Nije od krucijalnog značaja ali bilo bi lepo da na par mesta upotrebite neku reč koja je opisnija od one koja je nastala od reči "prepoznati". S jedne strane zato što u kontekstu u kom opisujete, baš je na nekim mestima generičko i ne daje puno informacija. S druge strane, u prvih par pasusa previše ponavljate ovu reč.

Motivacija projekta bila je u tome da se ne samo primene mnoge metode korišćene za prepoznavanje govora, već da se i uporede njihova praktičnost i tačnost. 

Primena prepoznavanja govora može se uočiti u mnogim svakodnevnim radnjama: audio pretraga na internetu, audio pretraga na uređajima za slepe ljude, pozivanje glasom, ...

Ovaj projekat se bavi prepoznavanjem konkretnih reči i njihova klasifikacija. Formulacija problema koji se rešava u ovom projektu se moze definisati na sledeći način: vrši se klasifikacija reči na jednu od 10 reči iz predodredjenog skupa. Ceo projekat rađen je u Python programskom jeziku.

> Projekat se zasniva na ideji korišćenja spektrograma kao osnovne metode prikaza zvuka u 2D formatu. Spektrogrami su korišćeni na dva načina tokom realizacije projekta. Jedan način podrazumeva korišćenje spektrograma u formi slike, čijom obradom možemo da izvučemo određene karakteristike iz zvuka. Drugi način podrazumeva ručno izvlačenje karakteristika iz spektrograma, koji je predstavljen matricom brojeva.

Reviewer 2: "Jedan način podrazumeva korišćenje spektrograma u formi slike, čijom obradom možemo da izvučemo određene karakteristike iz zvuka. Drugi način podrazumeva ručno izvlačenje karakteristika iz spektrograma, koji je predstavljen matricom brojeva."
Ove dve rečenice deluju kao ista stvar rečena na drugačiji način. Ja kontam šta ste hteli reći, ali, ako možete, malo preformulišite datu rečenicu tako da deo "... čijom obradom možemo da izvučemo određene karakteristike iz zvuka" ne zvuči kao da vi obrađujete ručno spektrogram i kao da znate koje su to karakteristike koje model uči iz njega.

Osvrt na rad ogleda se u setu metoda koje su pokrivene u referentnim radovima. Uloga ovih metoda može se podeliti u nekoliko kategorija: 

> 1. Izvlačenje karakteristika iz zvuka pomoću kepstralnih koeficijenata Mel skale (MFCC) 

Reviewer 2: ```suggestion
1. Izvlačenje karakteristika iz zvuka pomoću kepstralnih koeficijenata Mel skale (MFCC);
```

2. Klasifikatori, kojima su prosleđene MFCC karakteristike: Logistička regresija, Random Forest, SVM, XGBoost;

> 3. Konvolucione neuronske mreže (CNN) koje inkomponuju proces ekstrakcije karakteristika iz signala, kao i proces klasifikacije 

Reviewer 2: ```suggestion
3. Konvolucione neuronske mreže (CNN) koje inkomponuju proces ekstrakcije karakteristika iz signala, kao i proces klasifikacije.
```

### Metode

Rešenje datog problema prepoznavanja govora svodi se na izradu spektrograma i obradu istih.

#### Spektrogrami

Spektrogrami su vizuelne reprezentacije jačine signala. Mogu se posmatrati kao dvodimenzionalni grafici gde se može uočiti i treća dimenzija preko boja svakog dela spektrograma. Vremenska osa se gleda sa leve na desnu stranu po horizontalnoj osi. Vertikalna osa predstavlja frekvencijske komponente prisutne u signalu, dok boja označava jačinu svake od tih komponenti. U logaritamskoj je skali kako bi se prilagodila ljudskom uhu koje čuje po istom principu, što je dalje objašnjeno u samom radu.

Spektrogram služi za prikazivanje aplitude svake frekvencijske komponente signala u vremenskom intervalu. Intervali su mali da bi se moglo pretpostaviti da se amplitude frekvencijskih komponenti ne menjaju.

Boja na grafiku predstavlja amplitudu signala u određenom vremenskom trenutku. Plava boja na spektrogramu predstavlja niske amplitude, dok crvena boja predstavlja visoke amplitude.

> ![spektrogram](static\images\2.png)

Reviewer 2: Slika 2.png ne postoji u static\images segmentu.

#### Metode obrade spektrograma

##### 1. Logistička regresija

Logistička regresija je metoda klasifikacije koja se može primeniti i koristiti svuda gde imamo promenljive koje se mogu kategorisati. Za razliku od linearne regresije, vrednosti njenih rezultata su ograničene između 0 i 1. 

U slučaju binarne klasifikacije, ova metoda umesto linearne koristi sigmoidnu funkciju. U slučaju više klasa, koristi se softmax funkcija. Sigmoidna i softmax funckija prikazane su na slici:

![Sigmoid](static\images\Sigmoid.svg)

Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. U slučaju kada imamo više od dve klase, koristi se multinomijalna logistička regresija (Softmax Regression). Na sledećoj slici prikazana je Softmax funkcija.

![Softmax](static\images\Softmax.svg)

Kriterijumska funkcija ove metode je logaritamska kako bi se postiglo da gradient descent nađe globalni, a ne samo lokalni minimum funkcije.

$$\begin{equation}

> \sigma(\vec{z})_i=e^{z_i}*(\sum_{j=1}^{K}e^{z_j})^{-1} 

Reviewer 2: Nisam sigurna čemu ova strelica iznad vrednosti z. Ako hoćete da naglasite da je vektor, predstavite kao boldovan tekst.

Reviewer 2: Takođe, ako je to z vektor, onda ove dve strane ne čine jednakost. S desne strane je rezultat skalar.

Author: Imamo tu problem oko tog zapisa, neće da se lepo napiše ma kako stavili, pokazali smo vam već za taj slučaj, a ima i još nekoliko mesta

title: Prepoznavanje govora
summary: Projekat iz prepoznavanja govora rađen na letnjem kampu za stare polaznike 2022. godine od Dimitrija Pešića i Lazara Zubovića.
---
### Apstrakt

### Apstrakt na engleskom
### Uvod
Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. – Prepoznavanje govora je proces osposobljavanja nekog modela da prepozna i odreaguje na zvuk proizveden ljudskim govorom. Model uzima audio signal u formi talasa, izvlači iz njega podatke, obrađuje ih, prepoznaje i preduzima određene korake u zavisnosti od rezultata.

Motivacija projekta bila je u tome da se ne samo primene mnoge metode korišćene za speech recognition, već da se i uporede njihova praktičnost i tačnost. 

Primena prepoznavanja govora može se uočiti u mnogim svakodnevnim radnjama: audio pretraga na internetu, audio pretraga na uređajima za slepe ljude, voice dialing, ...

Ovaj projekat se bavi prepoznavanjem konkretnih reči i njihova klasifikacija – ograničen je na prepoznavanje i klasifikaciju svega deset reči. Ceo projekat rađen je u Python programskom jeziku.

> Projekat se zasniva na ideji korišćenja spektrograma kao osnovne metode prikaza zvuka u 2D formatu. Spektrogrami su korišćeni na dva načina tokom realizacije projekta. Može se koristiti kao slika čijom obradom dobijamo određene karakteristike zvuka, a u nekim metodama ne možemo koristiti kao sliku, već ručno moramo izvlačiti odlike zvuka.

Reviewer: ```suggestion
Projekat se zasniva na ideji korišćenja spektrograma kao osnovne metode prikaza zvuka u 2D formatu. Spektrogrami su korišćeni na dva načina tokom realizacije projekta. Jedan način podrazumeva korišćenje spektrograma u formi slike, čijom obradom možemo da izvučemo određene karakteristike iz zvuka. Drugi način podrazumeva ručno izvlačenje karakteristika iz spektrograma, koji je predstavljen matricom brojeva.```

Osvrt na rad ogleda se u setu metoda koje su pokrivene u referentnim radovima. Uloga ovih metoda može se podeliti u nekoliko kategorija: 

> 

Reviewer: Neka bude numerisano nabrajanje, ne crtice.

Reviewer 3: Rešeno, 165. linija

> -ekstrakcija odlika zvuka, što je posao za MFCC (Mel-Frequency Cepstral Coefficients); 

Reviewer: ```suggestion
1. Izvlačenje karakteristika iz zvuka pomoću kepstralnih koeficijenata Mel skale (MFCC) 
```
E sad u buduce mozete da pisete MFCC

-klasifikatori: logistička regresija, Random Forest, SVM, XGBoost;

> -kombinacija: upotreba konvolucionih neuronskih mreža (mreža samostalno uči koje odlike zvuka treba da izvuče, da bi ih samostalno i klasifikovala).

Reviewer: ```suggestion
3. Konvolucione neuronske mreže (CNN) koje inkomponuju proces ekstrakcije karakteristika iz signala, kao i proces klasifikacije 
```

> Do ovog projekta, metode su testirane na bazama podataka velikog kvaliteta i sa velikim brojem instanci. Nasuprot njihovim, ovaj rad ima dosta limitiranu bazu, te i sami rezultati variraju u odnosu na već dobijene.

Reviewer: Ovo nije tačno. Prvobitna baza koju ste koristili je solidne veličine, što se i ogleda u uspehu CNN-a na validaciji. Obrisao bih ceo ovaj pasos.

Reviewer 3: Rešeno, obrisano

Reviewer: Pa ne vidim bas da je reseno, pasos je i dalje tu

### Metode

Rešenje datog problema prepoznavanja govora svodi se na izradu spektrograma i obradu istih.

#### Spektrogrami

> Spektrogrami su vizuelne reprezentacije jačine signala. Mogu se posmatrati kao dvodimenzionalni grafici gde se može uočiti i treća dimenzija preko boja svakog dela spektrograma. Vremenska osa se gleda sa leve na desnu stranu po horizontalnoj osi. Vertikalna osa predstavlja frekvenciju koju možemo posmatrati i kao ton zvuka. U logaritamskoj je skali kako bi se prilagodila ljudskom uhu koje čuje po istom principu, što je dalje objašnjeno u samom radu.

Reviewer: ```suggestion
Spektrogrami su vizuelne reprezentacije jačine signala. Mogu se posmatrati kao dvodimenzionalni grafici gde se može uočiti i treća dimenzija preko boja svakog dela spektrograma. Vremenska osa se gleda sa leve na desnu stranu po horizontalnoj osi. Vertikalna osa predstavlja frekvencijske komponente prisutne u signalu, dok boja označava jačinu svake od tih komponenti. U logaritamskoj je skali kako bi se prilagodila ljudskom uhu koje čuje po istom principu, što je dalje objašnjeno u samom radu.
```

> Spektrogram služi za prikazivanje aplitude svake frekvencijske komponente signala u vremenskom intervalu. Intervali su mali da bi se moglo pretpostaviti da se amplitude frekvencijskih komponenti ne menjaju.

Reviewer: ```suggestion
Spektrogram služi za prikazivanje amplitude svake frekvencijske komponente signala u vremenskom intervalu. Intervali su mali, te se može pretpostaviti da se amplitude frekvencijskih komponenti ne menjaju u okviru jednog intervala.
```

> Boja na grafiku predstavlja amplitudu signala u određenom vremenskom trenutku. Plava boja na spektrogramu predstavlja niske amplitude, dok crvena boja predstavlja visoke amplitude.

Reviewer: Ovo koja boja sta znaci je nerelevantno ako ne prikazete sliku spektrograma gde je bas to tacno za boje. Ako ne postavite sliku, obrisite ceo ovaj pasos

Author: Rešeno

Reviewer 2: Slika 2.png ne postoji u static\images segmentu.

> Primena spektrograma u ovom radu jeste prepoznavanje fonema reči kako bi, spajanjem istih, reč mogla da se prepozna.

Reviewer: Ne radite ovako! Vi vrsite klasifikaciju na nivou reci, nema sklapanja fonema, to je bilo tacno u onom nekom starom radu za speech rec. koji ste predlozili, a koji ne koristimo

#### Metode obrade spektrograma

##### 1. Logistička regresija

Logistička regresija je metoda klasifikacije koja se može primeniti i koristiti svuda gde imamo promenljive koje se mogu kategorisati. Za razliku od linearne regresije, vrednosti njenih rezultata su ograničene između 0 i 1. 

> Ova metoda za binarnu klasifikaciju ne koristi linearnu već sigmoidnu funkciju bilo kog tipa, a softmax funkciju kada imamo slučaj sa više klasa.Primer sigmoidne funckije je dat na slici 1.

Reviewer: ```suggestion
U slučaju binarne klasifikacije, ova metoda umesto linearne koristi sigmoidnu funkciju. U slučaju više klasa, koristi se softmax funkcija. Sigmoidna funckija prikazana je na slici:
```

Reviewer 3: Rešeno

![Sigmoid](static\images\1.png)

> 

Reviewer: Dodajte softmax sliku isto, posto samo to i koristite

Reviewer 2: Dodajte labele za ose u svakoj od generisanih slika.

> Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. To često nije dovoljno, pa se koristi multinomijalna logistička regresija (ili Softmax Regression) koja može da razlikuje više od dve različite kategorije.

Reviewer: ```suggestion
Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. U slučaju kada imamo više od dve klase, koristi se multinomijalna logistička regresija (Softmax Regression) .
```

Ako su dobijeni i željeni rezultat istog znaka, vrednost kriterijumske funkcije je jednaka nuli, dok u suprotnom računamo gubitak. Na to moramo dodati i parametar za regularizaciju koji služi da izjednači uticaj maksimizacije granice i minimizacije gubitka.

$$
\min _w \lambda\|w\|^2+\sum_{i=1}^n\left(1-y_i\left\langle x_i, w\right\rangle\right)_{+}
$$

Nakon toga možemo izvesti gradijente za ažuriranje vrednosti težina modela:

$\frac{\delta}{\delta w_k} \lambda\|w\|^2=2 \lambda w_k \\$

$\frac{\delta}{\delta w_k}\left(1-y_i\left\langle x_i, w\right\rangle\right)_{+}= \begin{cases}0, & \text { if } y_i\left\langle x_i, w\right\rangle \geq 1 \\ -y_i x_{i k}, & \text { else }\end{cases}$

Težine ažuriramo zavisno od toga da li je naš klasifikator tačno klasifikovao novi podatak ili ne. Ukoliko jeste, ažuriramo samo gradijent regularizacionog parametra:

$$
w=w-\alpha \cdot(2 \lambda w)
$$

U suprotnom, ako je model napravio grešku, moramo da uključimo i funkciju gubitka u račun:

$$
w=w+\alpha \cdot\left(y_i \cdot x_i-2 \lambda w\right)
$$

##### 6. Konvolucione neuronske mreže

> Metoda konvolucionih neuronskih mreža pomaže za klasifikaciju podataka pomoću tehnike dubokog učenja. Neuronske mreže rade po principu čovečjeg mozga (odatle i naziv): dobija određene podatke koji se obično nalaze u formatu baze podataka, obrađuje ih i vraća rezultate. Kontrolom rezultata obrade podataka se ta mreža trenira. Ona uči na svojim greškama i poboljšava rezultate obrade.

Reviewer 2: ```suggestion
Metoda konvolucionih neuronskih mreža pomaže za klasifikaciju podataka pomoću tehnike dubokog učenja. Neuronske mreže su inspirisane neuronima i sinapsama u ljudskom mozgu. U konvolucionu neuralnu mrežu pohranjujemo ulazne podatke u vidu spektrograma, nakon čega se oni provlače kroz nekoliko slojeva konvolucije, sažimanja i potpuno povezanih slojeva. Izlaz iz ove mreže se koristi za proračunavanje vrednosti kriterijumske funkcije, na osnovu čega se ažuriraju parametri mreže. Ovaj postupak se potom iterativno ponavlja u cilju minimizacije greške modela.
```

Konvoluciona neuronska mreža korišćena u ovom projektu sastoji se iz 5 konvolucionih slojeva, koristi se 4 slojeva sažimanja, kao i 3 potpuno povezana sloja. 

Ceo proces može se svesti na sledeće korake: 

> - Spektrogram se prvo obrađuje konvolucijom i ReLU-om

Reviewer 2: ```suggestion
- Na ulaznu sliku se primenjuje dvodimenzionalnih konvolucija sa prethodno definisanim brojem kanala a potom i ReLU aktivaciona funkcija.
```

Reviewer 3: Rešeno

> - Smanjujemo veličinu obrađene slike pooling slojem

Reviewer 2: ```suggestion
- Smanjujemo veličinu obrađene slike slojem sažimanja (eng. _maximum pooling_).
```

Reviewer 3: Rešeno

> - Ponavljamo ovaj proces

Reviewer 2: Napomenuti da se ovaj proces ponavlja 4 puta i da poslednji sloj konvolucije nije praćen slojem sažimanja.

Author: ok

> Konvolucija (po čemu nastaje termin konvolucione neuronske mreže) u obradi slike je operator koji predstavlja sužavanje početne slike množenjem iste određenim filterom. 

Reviewer 2: U opštem slučaju ne sužava sliku pa samim tim to i nije informacija koju treba istaći.

Reviewer 3: Rešeno

Konvolucija kao bitne detalje posmatra one koji su mnogo puta uhvaćeni u kernelu. Problem može da se desi kada kernel ne zahvata ivice dosta puta, te može mnogo da smanji određenu sliku, a samim tim i da se reši ivičnih detalja. Ako do te pojave dođe, koristi se tehnika koja se zove padding. 

Padding označava dodavanje piksela na ivice. Samim tim, kada konvolucija radi svoj posao, ona će svojim kernelom mnogo puta pokriti tu površinu. 

ReLU (rectified linear activation function / rectified linear unit) je funkcija koja negativnim vrednostima daje nulu, a pozitivne ostavlja kakve jesu. Time dobijamo nelinearan model.

![Funkcija](static\images\fja.png)

Kroz neuronsku mrežu se propušta već napravljen spektrogram, kao i labele tih spektrograma koje mreža treba da prepozna.

> Za treniranje mreže koriste se dve metode simultano: loss funkcija i back propagation.

Reviewer 2: Funkcija greške odnosno loss funkcija nije metoda, to je funkcija. Dve metode koje se odvijaju jedna za drugom su propagacija unapred i propagacija unazad.

Author: Rešeno

Author: Rešeno

Reviewer 4: Kad resite, kliknite na _Resolve conversation_ jer bez toga nece moci da se zatvori PR i merguje grana.

> Cost funkcija (funkcija troškova ili gubitka) jeste funkcionalna veza željenog outputa i dobijenog outputa u funkciji, a loss funkcija je srednja vrednost svih cost funkcija.

Reviewer 2: Kriterijumska funkcija, ne funkcija troškova ili gubitka. Takođe, kriterijumska funkcija je usrednjena vrednost svih funkcija greške (loss function), nije obrnuto kao što ste vi napisali.

Author: Rešeno

Author: Rešeno

Najkorišćenija loss funkcija je Cross Entropy Loss. 

> Cross Entropy Loss radi tako što pokušava da minimizuje razliku između tačnih rezultata i verovatnoće predviđanja, to jest output.

Reviewer 2: Cilj je da se minimizuje greška unakrsne entropije. Netačno je reći da greška unakrsne entropije pokušava da bilo šta minimizuje.

Author: ok

Formula po kojoj se računa Cross Entropy Loss je sledeća:

> $H_p(q)=-\frac{1}{N} \sum_{i=1}^N y_i \cdot \log \left(p\left(y_i\right)\right)+\left(1-y_i\right) \cdot \log \left(1-p\left(y_i\right)\right)$

Reviewer 2: Ovo je binarna unakrsna entropija. Vama je neophodna forula za kategoričku unakrsnu entropiju (categorical cross entropy loss). Slične je forme ali postoje razlike.

Back propagacija je metod smanjenja grešaka u CNN posmatranjem neophodnih promena u prethodnom sloju od aktivacije da bi se u određenom sloju neuroni aktivirali na određen način.

> Backpropagation prolazi krroz sve primere i traži sumu svih težina veza među neuronima.

Reviewer 2: Ovaj i prethodni pasus nisu tačni. Niste baš najbolje razumeli algoritam propagacije unazad.

Reviewer 3: Promenjeno, proveri valja li

> Težine se menjaju u cilju računanja dovoljno dobrog gradient descenta za traženje lokalnog / maksimalnog minimuma ove funkcije, to jest tačnu reč.

Reviewer 2: Bolje je koristiti termin parametri mreže. 

Gradijentni spust je algoritam za minimizaciju greške modela. Ne tražimo mi dovoljno dobar gradijentni spust, težimo tome da gradijent kriterijumske funkcije bude blizak nuli.

Reviewer 3: Promenjeno, proveri valja li

![SGD](static/images/sgd.png)

### Istraživanje i rezultati

> Testiranje metoda vršeno je na dve baze: FSDD i srpske baze kreirane za potrebe projekta.

Reviewer: ```suggestion
Testiranje metoda vršeno je na dve baze: FSDD baze i baze srpskih reči, koja je kreirana za potrebe projekta.
```

> FSDD baza sadrži engleske cifre od 0 do 9 koje su izgovorene od strane 50 različitih ljudi. Sadrži ukupno 3000 snimaka.

Reviewer: Da li je od svake osobe isti broj snimaka, odnosno da li je baza balansirana? Dodajte komentar na tu temu

> Srpska baza sadrži 10 srpskih reči, gde su specifično birane reči koje su slične po nekim karakteristikama (ponavljanje slova, zamena slova, umanjenice, ...). Baza ukupno sadrži 500 snimaka, gde preko 10 ljudi izgovara ove reči različitim naglaskom, intonacijom i slično.

Reviewer: 1. Dajte (npr jedan) primer reci iz baze gde ukazujete na te slicnosti karakteristika
2. Recite kolko tacno ljudi, ovde treba biti precizan kad je vec broj tolko mali

Za konvolucionu neuronsku mrežu, potrebni su nam bili pokazatelji kako mreža uči tokom epoha treniranja. Baze su podeljene na trening, test i validacionu bazu, tako da je trening set sadržao 70% reči, a test i validacioni set po 15% reči u slučaju obe baze.

Rezultati su prikazani u tabeli ispod.

![Rezultati](static\images\4.png)

> Metrika ovih rezultata bila je tačnost, zato što je, zbog izbalansirane baza, ovo reprezentativna metrika.

Reviewer: 1. Podelite ovo na dve recenice
2. koristite izraz balansirane umesto izbalansirane

> Rezultati se dele po tome da li metoda koristi duboko učenje ili ne. Posmatrajući tabelu, konvoluciona neuronska mreža je ostvarila najveću tačnost kao metoda sa dubokim učenjem, a XGBoost daje najbolje rezultate među metodama koje ne koriste duboko učenje. Najmanje rezultate daje SVM sa polinomijalnim kernelom.

Reviewer: Prva recenica, ajde da kazemo da mozemo posmatrati odvojeno metode sa dubokim ucenjem i one bez (ovako deluje da sama tabela pravi tu podelu, sto nije slucaj)

Druga recenica pocetak: Iz tabele se moze uociti ... 
Poslednja recenica: Ovde cemo da primenimo onaj nas dogovor, pricamo samo o najboljem rezultatu SVM-a i ne diskutujemo o tome koji je kernel koriscen.

> U FSDD bazi dato je 10 labela, pa klasifikator radi odličan posao da pretpostavi u koju kategoriju labela određeni zvuk spada (cifra od 0 do 9).

Reviewer: Obrisao bih ovaj pasos. Nema razloga da 10 labela implicira dobre rezultate

> 

Reviewer: Okej, ovde nam treba malo duzi deo o matricama konfuzije. Evo kako to treba da izgleda:
Prikazane su matrice konfuzije na FSDD bazi za CNN, XGBoost, SVM, Random Forest i Logisticku regresiju. 
* slike matrica
Iz ovih matrica konfuzije, mozemo da zakljucimo sledece:
(Ovde vi treba da gledate matrice i pokusate da vidite neke pravilnosti ili zanimljivosti.
Evo jedna pomoc za pocetak: Primetite da je prisutan sablon da se zvuk jedinice cesto pogresno klasifikuje kao zvuk devetke. Diskutujte da li to ima nekog fizickog smisla, iz pogleda nacina izgovora "one" i "nine")

> U ovoj bazi podataka, određene reči mogu lako da se pomešaju na spektrogramu, pa su neke vrednosti vrlo blizu odlučnoj granici i da pomute labele. Iz tog razloga, rezultati SVM metode su veoma dobri.

Reviewer: Ako dodate ovaj prethodni deo, ovaj pasos nije mnogo potreban

> Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Deep learning metode povoljnije su za feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao speech recognition loss funkcija pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~200Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.

Reviewer: 2. rec: DL metode same vrse feature extraction proces
Ovo ostalo ste lepo rekli, super

> Gledajući ova dva faktora u obzir, očekivano je da će performansa CNN-a biti najbolja.

Reviewer: Nije deo za rezultate mesto za pisanje hipoteza -> brisi

Rezultati koji su odađeni na srpskoj bazi podataka dosta su slabiji u poređenju sa engleskom bazom. Srpska baza pravljena je u amaterskim uslovima: mikrofon slabijeg kvaliteta, dosta šuma se može čuti u samim snimcima, nisu svi zvuci iste jačine, kao ni dužine. Ovi faktori dosta utiču na kvalitet spektrograma, na kome ima dosta više šuma u poređenju sa spektrogramom engleske baze.

Rezultate vizuelno možemo prikazati matricama konfuzije. 

![Rezultati](static/images/LinearSVM.png)

![Rezultati](static/images/LogisticRegression.png)

![Rezultati](static/images/RandomForest.png)

> ![Rezultati](static/images/XGB.png)

Reviewer: Sad sledi oopeet jedna diskusija o tome sta vidimo na ovim matricama konfuzije.

### Zaključak

> Projekat koristi FSDD bazu podataka za poređenje performansi pri prepoznavanju govora između sledećih metoda: SVM, MFCCs, CNN, Random Forest, XGBoost i logistička regresija. Uz ovu i samostalno napravljenu srpsku bazu podataka, ove metode su se pokazale kao veoma uspešne pri detektovanju izgovorenih reči. CNN model je imao najveću uspešnost pri prevođenju reči, a SVM sa RBF kernelom najmanju. Tačnost između metoda varira od 51.45% do 97.28%, pa je zaključak ovog rada da je tačnost CNN modela značajno veća od ostalih testiranih modela.

Reviewer: 1. Prva recenica zakljucka treba da kaze sta je projekat, a posle recite kakva baza postoji
2. Iskreno nikog ne zanima najmanja postignuta tacnost, jer mogu ja dodjem kazem moja metoda lupanja random daje 10% tacnosti (mos' misliti), tako da to mozete obrisati
3. Nije zakljucak rada da je CNN bolji, to pise u tabeli. Zakljucak je sta mozete zakljuciti iz toga i npr kakve to implikacije moze da ima ka daljem radu

title: Prepoznavanje govora
summary: Projekat iz prepoznavanja govora rađen na letnjem kampu za stare polaznike 2022. godine od Dimitrija Pešića i Lazara Zubovića.
---
### Apstrakt

> Prepoznavanje govora predstavlja jedan od najvećih izazova tehnologije. Sve veća potreba za digitalizacijom dovodi do potrebom za širenjem znanja u ovom polju. Dosadašnja istraživanja pokazuju efikasnost i tačnost prepoznavanja govora mnogih metoda sa i bez korišćenja dubokog učenja, ne fokusirajući se toliko na pojašnjenja razlika u rezultatima koji su dobijeni. Ovaj rad se fokusira na posmatranju i upoređivanju metoda poput konvolucionih neuronskih mreža i raznih klasifikatora podataka koji ne koriste tehniku dubokog učenja kako bi se utvrdilo šta je najbolji pristup za identifikovanje reči. Testirajući na FSDD bazi reči i bazi podataka koja se sastoji od srpskih reči, utvrđeno je da najprecizniji način za obradu audio zapisa konvoluciona neuronska mreža, pa je najoptimalnije dalja istraživanja voditi u tom smeru.

Reviewer 2: "Dosadašnja istraživanja pokazuju efikasnost i tačnost prepoznavanja govora mnogih metoda sa i bez korišćenja dubokog učenja, ne fokusirajući se toliko na pojašnjenja razlika u rezultatima koji su dobijeni." Ovaj deo oko toga da se niko do sad nije fokusirao na poređenje rezultata nije tačno.

> Prepoznavanje govora predstavlja jedan od najvećih izazova tehnologije. Sve veća potreba za digitalizacijom dovodi do potrebom za širenjem znanja u ovom polju. Dosadašnja istraživanja pokazuju efikasnost i tačnost prepoznavanja govora mnogih metoda sa i bez korišćenja dubokog učenja, ne fokusirajući se toliko na pojašnjenja razlika u rezultatima koji su dobijeni. Ovaj rad se fokusira na posmatranju i upoređivanju metoda poput konvolucionih neuronskih mreža i raznih klasifikatora podataka koji ne koriste tehniku dubokog učenja kako bi se utvrdilo šta je najbolji pristup za identifikovanje reči. Testirajući na FSDD bazi reči i bazi podataka koja se sastoji od srpskih reči, utvrđeno je da najprecizniji način za obradu audio zapisa konvoluciona neuronska mreža, pa je najoptimalnije dalja istraživanja voditi u tom smeru.

Reviewer 2: ```suggestion
Prepoznavanje govora predstavlja jedan od najvećih izazova tehnologije. Sve veća potreba za digitalizacijom dovodi do potrebom za širenjem znanja u ovom polju. Dosadašnja istraživanja pokazuju efikasnost i tačnost prepoznavanja govora mnogih metoda sa i bez korišćenja dubokog učenja, ne fokusirajući se toliko na pojašnjenja razlika u rezultatima koji su dobijeni. Ovaj rad se fokusira na posmatranje i upoređivanje metoda poput konvolucionih neuronskih mreža, kao i nekoliko klasifikatora podataka koji ne koriste tehniku dubokog učenja, kako bi se utvrdilo šta je najbolji pristup za identifikovanje reči. Testirajući modele na FSDD bazi reči i bazi podataka koja se sastoji od srpskih reči, utvrđeno je da najtačnije rezultate pri obradi audio zapisa donosi konvoluciona neuronska mreža. Iz ovoga zaključujemo da je optimalno dalja istraživanja usmeriti ka dubokom učenju.
```

### Apstrakt na engleskom
### Uvod
Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. Prepoznavanje govora je proces osposobljavanja nekog modela da identifikuje i odreaguje na zvuk proizveden ljudskim govorom. Model uzima audio signal u formi talasa, izvlači iz njega podatke, obrađuje ih i identifikuje izgovorenu reč.

Motivacija projekta bila je u tome da se ne samo primene mnoge metode korišćene za prepoznavanje govora, već da se i uporede njihova praktičnost i tačnost. 

> Primena projekta može se uočiti u mnogim svakodnevnim radnjama: audio pretraga na internetu, audio pretraga na uređajima za slepe ljude, pozivanje glasom, ...

Reviewer 2: ```suggestion
Primena projekta može se uočiti u mnogim svakodnevnim radnjama: audio pretraga na internetu, audio pretraga na uređajima za slepe ljude, pozivanje glasom, i slično.
```

> Ovaj projekat se bavi raspoznavanjem konkretnih reči i njihova klasifikacija. Formulacija problema koji se rešava u ovom projektu se moze definisati na sledeći način: vrši se klasifikacija reči na jednu od 10 reči iz predodredjenog skupa. Ceo projekat rađen je u Python programskom jeziku.

Reviewer 2: ```suggestion
Ovaj projekat se bavi raspoznavanjem konkretnih reči i njihovom klasifikacijom. Formulacija problema koji se rešava u ovom projektu se moze definisati na sledeći način: vrši se klasifikacija reči na jednu od 10 reči iz predodredjenog skupa. Ceo projekat rađen je u Python programskom jeziku.
```

> Projekat se zasniva na ideji korišćenja spektrograma kao osnovne metode prikaza zvuka u 2D formatu. Spektrogrami su korišćeni na dva načina tokom realizacije projekta. Jedan način podrazumeva korišćenje spektrograma u formi slike gde korišćeni klasifikatori obrađuju zvuk i pronalaze određene karakteristike. Drugi način podrazumeva ručno izvlačenje karakteristika iz spektrograma, koji je predstavljen matricom brojeva.

Reviewer 2: Klasifikatori (CNN) ne obrađuju zvuk, oni donose zaključke o zvuku na osnovu slika.

Osvrt na rad ogleda se u setu metoda koje su pokrivene u referentnim radovima. Uloga ovih metoda može se podeliti u nekoliko kategorija: 

1. Izvlačenje karakteristika iz zvuka pomoću kepstralnih koeficijenata Mel skale (MFCC);

2. Klasifikatori, kojima su prosleđene MFCC karakteristike: Logistička regresija, Random Forest, SVM, XGBoost;

3. Konvolucione neuronske mreže (CNN) koje inkomponuju proces ekstrakcije karakteristika iz signala, kao i proces klasifikacije.

### Metode

Rešenje datog problema prepoznavanja govora svodi se na izradu spektrograma i obradu istih.

#### Spektrogrami

Spektrogrami su vizuelne reprezentacije jačine signala. Mogu se posmatrati kao dvodimenzionalni grafici gde se može uočiti i treća dimenzija preko boja svakog dela spektrograma. Vremenska osa se gleda sa leve na desnu stranu po horizontalnoj osi. Vertikalna osa predstavlja frekvencijske komponente prisutne u signalu, dok boja označava jačinu svake od tih komponenti. U logaritamskoj je skali kako bi se prilagodila ljudskom uhu koje čuje po istom principu, što je dalje objašnjeno u samom radu.

> Spektrogram služi za prikazivanje aplitude svake frekvencijske komponente signala u vremenskom intervalu. Intervali su mali, te se može pretpostaviti da se amplitude frekvencijskih komponenti ne menjaju u okviru jednog intervala.

Reviewer 2: ```suggestion
Spektrogram služi za prikazivanje amplitude svake frekvencijske komponente signala u nekom vremenskom intervalu. Intervali su mali, te se može pretpostaviti da se amplitude frekvencijskih komponenti ne menjaju u okviru jednog intervala.
```

#### Metode obrade spektrograma

##### 1. Logistička regresija

Logistička regresija je metoda klasifikacije koja se može primeniti i koristiti svuda gde imamo promenljive koje se mogu kategorisati. Za razliku od linearne regresije, vrednosti njenih rezultata su ograničene između 0 i 1. 

> U slučaju binarne klasifikacije, ova metoda umesto linearne koristi sigmoidnu funkciju. U slučaju više klasa, koristi se softmax funkcija. Sigmoidna i softmax funckija prikazane su na slici:

Reviewer 2: ```suggestion
U slučaju binarne klasifikacije, ova metoda umesto linearne koristi sigmoidnu funkciju. U slučaju više klasa, koristi se softmax funkcija. Sigmoidna funkcija prikazana je na slici:
```

![Sigmoid](static\images\Sigmoid.svg)

> Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. U slučaju kada imamo više od dve klase, koristi se multinomijalna logistička regresija (Softmax Regression). Na sledećoj slici prikazana je Softmax funkcija.

Reviewer 2: ```suggestion
Binarna logistička regresija kao izlaz daje vrednosti 0 ili 1, zavisno od toga da li posmatrana promenljiva pripada nekoj klasi ili ne. U slučaju kada imamo više od dve klase, koristi se multinomijalna logistička regresija (Softmax Regression).
```

> Kriterijumska funkcija ove metode je logaritamska kako bi se postiglo da gradient descent nađe globalni, a ne samo lokalni minimum funkcije.

Reviewer 2: Ova rečenica je redundantna u kontekstu toga da se ne može mnogo zaključiti iz nje. Ili dajte malo bolje pojašnjenje rečenice ili je uklonite.

$$\begin{equation}

> \sigma(z)_i=e^{z_i}*(\sum_{j=1}^{K}e^{z_j})^{-1} 

Reviewer 2: ```suggestion
\sigma(z_{i})=e^{z_i}*(\sum_{j=1}^{K}e^{z_j})^{-1} 
```

> \sigma(z)_i=e^{z_i}*(\sum_{j=1}^{K}e^{z_j})^{-1} 

Reviewer 2: Napišite dodatnu rečenicu koja opisuje ovu jednačinu, odnosno aktivacionu funkciju.

\end{equation}$$

Postoji slučaj kada nam se izbor svodi na dve kategorije. Da bi logistička regresija dala što bolje rezultate, trenira se MLE (Maximum Likelihood Estimation) metodom. Pomoću ove metode dobijamo verovatnoće za svaki primer, pa se one logaritmuju i sabiraju i time formiraju konačnu predviđenu verovatnoću. Svaka vrednost iznad 0.5 (ili bilo koje zadate granice) se tretira kao da je jedinica, a svaka manja od te granice se tretira kao nula.

U slučaju kada imamo više kategorija (u našem slučaju 10), koristi se Softmax regresija umesto Sigmoida kako bismo dobili deset verovatnoća čija je suma 1. Konačnu odluku o pravom izboru donosimo po tome koja kategorija ima najveću verovatnoću za zadate ulazne podatke.

##### 2. MFCCs

> MFCCs (Mel-Frequency Cepstral Coefficients) jesu koeficijenti koji opisuju karakteristike na osnovu spektrograma određenog zvuka. Njihova primena u ovom projketu svodi se na izdvajanje ključnih odlika nekog zvuka kako bi reč mogla da se prepozna. Te odlike se zovu formonti i njih stvara ljudski vokalni trakt prilikom govora, menjajući čist glas koji stvaraju naše glasne žice dok vibriraju. Ove odlike se formiraju u reč.

Reviewer 2: ```suggestion
MFCCs (Mel-Frequency Cepstral Coefficients) jesu koeficijenti koji opisuju karakteristike zvuka na osnovu njegovog spektrograma. Njihova primena u ovom projketu svodi se na izdvajanje ključnih odlika nekog zvuka kako bi reč mogla da se prepozna. Te odlike se zovu formonti i njih stvara ljudski vokalni trakt prilikom govora, menjajući čist glas koji stvaraju naše glasne žice dok vibriraju. Ove odlike se formiraju u reč.
```

Kepstar (cepstrum) se može intuitivno predstaviti kao spektar spektra. On nastaje inverznom Furijeovom transformacijom logaritmovanog spektra. Formula za nastanak kepstra:

$$ C(x(t))=F^{-1}[\log (F[x(t)])] $$

Proces stvaranja kepstra je sledeći:

> 1. Na dobijeni signal primenimo diskretnu Furijeovu transformaciju. Ova transformacija nam daje funkciju zavisnosti jačine zvuka od frekvencije po sledećoj formuli:

Reviewer 2: ```suggestion
1. Na dobijeni signal primenimo diskretnu Furijeovu transformaciju:
```

$$ \begin{aligned} X_k &=\sum_{n=0}^{N-1} x_n \cdot e^{-\frac{i 2 \pi}{N} k n} \\ &=\sum_{n=0}^{N-1} x_n \cdot\left[\cos \left(\frac{2 \pi}{N} k n\right)-i \cdot \sin \left(\frac{2 \pi}{N} k n\right)\right] \end{aligned} $$

> Grafik koji dobijemo ovom formulom zove se spektar snage. Spektar snage je pokazatelj amplituda svih sinusoida određenog zvuka u odnosu na frekvenciju tih sinusoida.

Reviewer 2: Kvadriranjem modula spektra dobijamo spektar snage.

Reviewer 2: ```suggestion
Kvadriranjem amplitudskog spektra dobijamo spektar snage.
```

> Grafik koji dobijemo ovom formulom zove se spektar snage. Spektar snage je pokazatelj amplituda svih sinusoida određenog zvuka u odnosu na frekvenciju tih sinusoida.

Reviewer 2: "Spektar snage je pokazatelj amplituda svih sinusoida određenog zvuka u odnosu na frekvenciju tih sinusoida."

Ovo nema smisla. Slobodno obrišite.

$$L_{C E}(\hat{y}, y)=-[y \log \sigma(\mathbf{w} \cdot \mathbf{x}+b)+(1-y) \log (1-\sigma(\mathbf{w} \cdot \mathbf{x}+b))]$$

Propagacija unazad je metod smanjenja grešaka u CNN posmatranjem neophodnih promena vrednosti parametara mreže u svakom sloju kako bi se neuroni aktivirali na određen način.

Propagacija unazad prolazi kroz sve slojeve i menja parametre mreže u svakom koraku.

Parametri mreže se menjaju u cilju računanja dovoljno dobrog gradient spusta za traženje lokalnog / maksimalnog minimuma ove funkcije. Dakle, teži se tome da gradijent kriterijumske funkcije bude što bliži nuli.

![SGD](static/images/Backpropagation.svg)

### Istraživanje i rezultati

Testiranje metoda vršeno je na dve baze: FSDD baze i baze srpskih reči, koja je kreirana za potrebe projekta.

FSDD baza sadrži engleske cifre od 0 do 9 koje su izgovorene od strane 50 različitih ljudi. Sadrži ukupno 3000 snimaka.

Srpska baza sadrži 10 srpskih reči, gde su specifično birane reči koje su slične po nekim karakteristikama (ponavljanje slova, zamena slova, umanjenice, ...). Baza ukupno sadrži 500 snimaka, gde je 29 ljudi izgovaralo ove reči različitim naglaskom i intonacijom.

> U FSDD bazi podataka, 6 osoba je izgovorila svaku reč 50 puta, kako bi ukupno bilo 3000 snimaka, što čini vrlo balansiranu bazu podataka. U srpskoj bazi, 10 reči je rečeno od strane 27 ljudi, dok su dve osobe ponovile izgovaranje ovih 10 reči 13 i 10 puta. Njihovih snimaka je 130 i 100, pa otuda i 500 snimaka u bazi. Korišćeni su drugačiji izgovori i intonacije zbog raznovrsnosti. Baza je slabije balansirana, što se odrazuje na same rezultate testiranja.

Reviewer: "Baza je slabije balansirana, što se odrazuje na same rezultate testiranja." 
Baza je balansirana ako ima maltene isto pojavljivanja svake klase za klasifikaciju. To sto su dve osobe vise puta govorile ne znaci da nije balansirana, i dalje imate oko 50 izgovora svake od reci zar ne? Umesto manje balansirana, moguce je reci pristrasna. Ja sad prvi put cujem da ste ovo radili, bas je lose da na svesno napravite bazu pristrasnom, samo da biste dosli do "lepog broja" kao 500

> U bazi srpskih reči, u uređenoj trojci gluva - glava - plava očekuju se češće greške pri klasifikaciji. To se može očekivati jer su drugi i poslednja dva glasa isti. Takođe, kako su „P“ i „G“ oba praskavi suglasnici, to jest isti su po mestu tvorbe, veća je verovatnoća pojavljivanja greške.

Reviewer: nije zamerka samo pohvala sto ste lepo objasnili fonetsku pojavu do koje dolazi

Za konvolucionu neuronsku mrežu, potrebni su nam bili pokazatelji kako mreža uči tokom epoha treniranja. Baze su podeljene na trening, test i validacionu bazu, tako da je trening set sadržao 70% reči, a test i validacioni set po 15% reči u slučaju obe baze.

Rezultati su prikazani u tabeli ispod.

![Rezultati](static\images\Tabela.svg)

> Metrika ovih rezultata bila je tačnost. Zbog balansirane baze, ovo predstavlja zaista reprezentativnu metriku.

Reviewer: ```suggestion
Metrika ovih rezultata bila je tačnost. Zbog balansiranosti baze, ovo predstavlja zaista reprezentativnu metriku.
```

Odvojeno možemo posmatrati rezultate metoda sa dubokim učenjem i one bez dubokog učenja. Iz tabele se može uočiti da je konvoluciona neuronska mreža ostvarila najveću tačnost kao metoda sa dubokim učenjem, a XGBoost daje najbolje rezultate među metodama koje ne koriste duboko učenje.

> Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Metode sa dubokim učenjem same vrše feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao loss funkcija za prepoznavanje govora pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~200Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.

Reviewer: ```suggestion
Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Metode sa dubokim učenjem same vrše feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao loss funkcija za prepoznavanje govora pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~100Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.
```

> Rezultati koji su odađeni na srpskoj bazi podataka dosta su slabiji u poređenju sa engleskom bazom. Srpska baza pravljena je u amaterskim uslovima: mikrofon slabijeg kvaliteta, dosta šuma se može čuti u samim snimcima, nisu svi zvuci iste jačine, kao ni dužine. Ovi faktori dosta utiču na kvalitet spektrograma, na kome ima dosta više šuma u poređenju sa spektrogramom engleske baze.

Reviewer: ```suggestion
Tačnosti postignute na srpskoj bazi podataka značajno su niže u poređenju sa engleskom bazom. Srpska baza pravljena je u amaterskim uslovima: mikrofon slabijeg kvaliteta, dosta šuma se može čuti u samim snimcima, nisu svi zvuci iste jačine, kao ni dužine. Ovi faktori dosta utiču na kvalitet spektrograma, na kome ima dosta više šuma u poređenju sa spektrogramom engleske baze.
```

Rezultate vizuelno možemo prikazati matricama konfuzije. 

![Rezultati](static/images/XGB.svg)

![Rezultati](static/images/SVM1.svg)

![Rezultati](static/images/RandomForest.svg)

![Rezultati](static/images/LogistickaRegresija.svg)

Prikazane su matrice konfuzije na FSDD bazi za XGBoost, SVM, Random Forest i logističku regresiju, tim redosledom.

> Iz ovih matrica konfutije može se primetiti kako, ma koja se metoda koristi, brojevi dva, tri i četiri uvek imaju najveću tačnost pronalaženja. 

Reviewer: ```suggestion
Iz ovih matrica konfuzije može se primetiti kako, nezavisno od metode koja se koristi, cifre dva, tri i četiri uvek imaju najveću tačnost pronalaženja. 
```

> Brojevi 9 i 1 su često mešani pri klasifikaciji ova četiri modela. Njihov izgovor može se protumačiti kao sličan ("one" i "nine") pa su ova dva broja par sa najvećim sličnostima u karakteristikama.

Reviewer: ```suggestion
Cifre 9 i 1 su često mešane pri klasifikaciji kod ova četiri modela. Njihov izgovor se može protumačiti kao sličan ("one" i "nine"), te su ova dva broja par sa najvećim sličnostima u karakteristikama.
```

> U svim metodama, broj 6 se pokazao kao najteže klasifikovan i u svim metodama mešan je sa brojevima 3 i 8, što ima manje fizičkog smisla od brojeva 1 i 9 ("six","three","eight").

Reviewer: ```suggestion
U svim metodama, cifra 6 je najviše puta pogrešno klasifikovana. Najčešće je mešana sa ciframa 3 i 8, što ima manje fizičkog smisla od mešanja cifara 1 i 9 ("six","three","eight").
```

> XGBoost i Random Forest su se pokazale kao najbolje metoda koje ne koriste tehniku dubokog učenja. Bolje rezultate dobijamo zato što su Decision tree algoritmi jednostavni za implementaciju i zaista precizni. Mogu se dobiti odlični rezultati u minimizaciji greški u prepoznavanju korišćenjem tehnike stabla odlučivanja pri obradi karakteristika zvuka. Gledajući u tabelu, XGBoost je imao veću preciznost od Random Forest klasifikatora. Ovo se može objasniti "obrezivanjem drveća" koje XGBoost radi, to jest Boosting kojim poboljšava klasifikaciju. XGBoost pokušava da smanji kriterijumsku funkciju modela koji obrađuje podatke, dok se Random Forest ne fokusira na te parametre i ne optimizuje model tako da to odgovara loše balansiranoj bazi podataka. 

Reviewer: "Bolje rezultate dobijamo zato što su Decision tree algoritmi jednostavni za implementaciju i zaista precizni. Mogu se dobiti odlični rezultati u minimizaciji greški u prepoznavanju korišćenjem tehnike stabla odlučivanja pri obradi karakteristika zvuka." 

Ove recenice nemaju smisla na vise nivoa. Umesto ovoga rekao bih nesto kao: Obe navedene metode su ansambl metode koje koriste stabla odlučivanja, te su generalno veoma otporne na preprilagođavanje.

> XGBoost i Random Forest su se pokazale kao najbolje metoda koje ne koriste tehniku dubokog učenja. Bolje rezultate dobijamo zato što su Decision tree algoritmi jednostavni za implementaciju i zaista precizni. Mogu se dobiti odlični rezultati u minimizaciji greški u prepoznavanju korišćenjem tehnike stabla odlučivanja pri obradi karakteristika zvuka. Gledajući u tabelu, XGBoost je imao veću preciznost od Random Forest klasifikatora. Ovo se može objasniti "obrezivanjem drveća" koje XGBoost radi, to jest Boosting kojim poboljšava klasifikaciju. XGBoost pokušava da smanji kriterijumsku funkciju modela koji obrađuje podatke, dok se Random Forest ne fokusira na te parametre i ne optimizuje model tako da to odgovara loše balansiranoj bazi podataka. 

Reviewer: "Ovo se može objasniti "obrezivanjem drveća" koje XGBoost radi, to jest Boosting kojim poboljšava klasifikaciju. XGBoost pokušava da smanji kriterijumsku funkciju modela koji obrađuje podatke, dok se Random Forest ne fokusira na te parametre i ne optimizuje model tako da to odgovara loše balansiranoj bazi podataka."

Pazite ovde. XGBoost je isto prilično okej s nebalansiranom bazom. Kao prednost XGBoosta bih mzd samo naveo kako u svakoj iteraciji XGBoost pokusava da kompenzuje rezultate dosadasnjeg modela, te je bolji u prilagodjavanju trening podacima

> Konvoluciona neuronska mreža, kao metoda koja koristi duboko učenje, prevazišla je rezultate običnih metoda. To se može objasniti time što je CNN kompleksniji model, pa može da modeluje kompleksniju relaciju između paramatara koji su mu dati. Metode sa dubokim učenjem imaju bolju primenu u ovoj oblasti tehnologije zbog sličnosti ovih algoritama ljudskom mozgu. Jednostavnost struktura algoritama je glavna mana u primeni mašinskog učenja za klasifikovanje zvuka, kao i to što je neophodna veća intervencija čoveka pri podešavanjima algoritama i metoda. 

Reviewer: ```suggestion
Konvoluciona neuronska mreža, kao metoda koja koristi duboko učenje, prevazišla je rezultate običnih metoda. To se može objasniti time što je CNN kompleksniji model, pa može da modeluje kompleksniju relaciju između paramatara koji su mu dati. Metode sa dubokim učenjem imaju široku primenu u oblasti mašinskog učenja zbog sličnosti ovih algoritama ljudskom mozgu. Jednostavnije metode često ne uspevaju da modeluju kompleksne veze između podataka, te je neophodno odlučiti se za kompleksnije metode.
```

### Zaključak

> Projekat "Prepoznavanje govora" pokazuje načine rešavanja popularne dileme pretvaranja glasa u kucani tekst. Koristi se FSDD baza podataka za poređenje performansi pri prepoznavanju govora između sledećih metoda: SVM, MFCCs, CNN, Random Forest, XGBoost i logistička regresija. Uz FSDD, koristi se i samostalno napravljena baza podataka koja se sastoji od srpskih reči, gde je dokazano, testiranjem metoda, da su se ove metode pokazale kao veoma uspešne pri detektovanju izgovorenih reči. CNN model je imao najveću uspešnost pri prevođenju reči. Tačnost između metoda varira od 51.45% do 97.28%, pa je zaključak ovog rada da je u ovoj oblasti AI tehnologije, zbog lakoće snalaženja sa ogromnom količinom podataka i smanjivanjem broja parametara bez gubljenja bitnih informacija, CNN najpraktičnija metoda za rad, što znači da se dalja istraživanja mogu usmeravati u primeni ove metode.

Reviewer: ```suggestion
Projekat "Prepoznavanje govora" pokazuje načine rešavanja popularne dileme pretvaranja glasa u kucani tekst. Koristi se FSDD baza podataka za poređenje performansi pri prepoznavanju govora između sledećih metoda: SVM, CNN, Random Forest, XGBoost i logistička regresija. Uz FSDD, koristi se i samostalno napravljena baza podataka koja se sastoji od srpskih reči, gde je dokazano, testiranjem metoda, da su se ove metode pokazale kao veoma uspešne pri detektovanju izgovorenih reči. CNN model je imao najveću uspešnost pri prevođenju reči. Tačnost metoda dolazi čak do 97.28%, te je zaključak ovog rada da je CNN najpraktičnija metoda za rad. Dalja istraživanja bi trebalo usmeravati ka ispitivanju ove metode.
```

Najkorišćenija loss funkcija je Cross Entropy Loss. 
Cross Entropy Loss radi tako što pokušava da minimizuje razliku između tačnih rezultata i verovatnoće predviđanja, to jest output.

Formula po kojoj se računa Cross Entropy Loss je sledeća:

$L_{C E}(\hat{y}, y)=-[y \log \sigma(\mathbf{w} \cdot \mathbf{x}+b)+(1-y) \log (1-\sigma(\mathbf{w} \cdot \mathbf{x}+b))]$

XGBoost se u Pythonu implementira bibliotekom xgboost. 

##### 5. SVM

Posao SVM klasifikatora je da u N-dimenzionalnom prostoru, gde je N broj parametara, pronađe hiperravan koja na najbolji način klasifikuje sve tačke koje predstavljaju podaci.

Kako postoji velik broj ovih hiperravni, kao optimalnu uzimamo onu kod koje je udaljenost granice odlučivanja podjednako udaljena od podataka svih tipova. Ovim dobijamo veću verovatnoću da će bilo koji naknadno dodati podatak biti pravilno klasifikovan. 

Hiperravni koje ograničavaju zonu udaljenosti od granice odlučivanja na kojoj klasifikator daje vrednosti čija je apsolutna vrednost manja od 1 nazivaju se noseći vektori. To znači da za svaki podatak koji se nalazi unutar tih vektora ne možemo sa sigurnošću reći kojoj klasi pripada.

> ![SVM1](static/images/SVM.svg)

Reviewer 5: Neke stvari koje mi se ne sviđaju kod ovog grafika:
* Mnogo različitih debljina linija. Probajte da se držite jedne debljine, eventualno nešto podebljate/stanjite da bi dodali na značaju
* Veličine kružića i kvadratića su prevelike
* Font je mnogo mali
* Debljina linija na osama je preveliki
* Ove linije koje ste obeležili sa "vektori podrške" nisu vektori podrške. Prvo, označili ste dve linije kao "vektor", a linije nisu vektori. Drugo, "vektor podrške" (pretpostavljam da mislite na support vector) je normalan na hipotezu (liniju koja vrši spearaciju klasa). Linije možete nazvati "margine", ali proverite sa Natašom i Pađinom.
* Margine (isprekidane linije) treba da prolaze kroz sample-ove, a ne pored njih. Dakle, treba da prolaze kroz centre kružića i kvadratića

Evo dobrih primera SVM objašnjenja:
https://upload.wikimedia.org/wikipedia/commons/thumb/7/72/SVM_margin.png/1280px-SVM_margin.png
https://static.javatpoint.com/tutorial/machine-learning/images/support-vector-machine-algorithm.png
https://miro.medium.com/max/720/0*rIGWbxfCPvxr_cz1

U FSDD bazi podataka, 6 osoba je izgovorila svaku reč 50 puta, kako bi ukupno bilo 3000 snimaka, što čini vrlo balansiranu bazu podataka. U srpskoj bazi, 10 reči je rečeno od strane 27 ljudi, dok su dve osobe ponovile izgovaranje ovih 10 reči 13 i 10 puta. Njihovih snimaka je 130 i 100, pa otuda i 500 snimaka u bazi. Korišćeni su drugačiji izgovori i intonacije zbog raznovrsnosti. Baza je slabije pristrasna, što se odražava na same rezultate testiranja.

U bazi srpskih reči, u uređenoj trojci gluva - glava - plava očekuju se češće greške pri klasifikaciji. To se može očekivati jer su drugi i poslednja dva glasa isti. Takođe, kako su „P“ i „G“ oba praskavi suglasnici, to jest isti su po mestu tvorbe, veća je verovatnoća pojavljivanja greške.

Za konvolucionu neuronsku mrežu, potrebni su nam bili pokazatelji kako mreža uči tokom epoha treniranja. Baze su podeljene na trening, test i validacionu bazu, tako da je trening set sadržao 70% reči, a test i validacioni set po 15% reči u slučaju obe baze.

Rezultati su prikazani u tabeli ispod.

> ![Rezultati](static\images\Tabela.svg)

Reviewer 5: Zamenite .svg tabele sa markdown tabelom. Na taj način možemo da biramo stil tabela kasnije.
Pogledajte uputstvo za kako to da uradite, izguglajte ako vam treba još detalja ili pingujte Luku ako vam idalje nešto ne valja.

Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Metode sa dubokim učenjem same vrše feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao loss funkcija za prepoznavanje govora pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~100Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.

Tačnosti postignute na srpskoj bazi podataka značajno su niže u poređenju sa engleskom bazom. Srpska baza pravljena je u amaterskim uslovima: mikrofon slabijeg kvaliteta, dosta šuma se može čuti u samim snimcima, nisu svi zvuci iste jačine, kao ni dužine. Ovi faktori dosta utiču na kvalitet spektrograma, na kome ima dosta više šuma u poređenju sa spektrogramom engleske baze.

Rezultate vizuelno možemo prikazati matricama konfuzije. 

![Rezultati](static/images/XGB.svg)

![Rezultati](static/images/SVM1.svg)

> ![Rezultati](static/images/RandomForest.svg)

Reviewer 5: Ovo je hitno: Povećajte font
Ovo je manje hitno: Stavite da su pretpostavke na slici $p_1$, $p_2$, ... $p_n$ i da je konačna pretpostavka formula koja na osnovu ovih $p$-ova računa pretpostavku (nešto poput $\text{argmax}\_{k\in[0, K]} \sum_{p_i=k}1$)

Napomena: poslednji sloj konvolucije nije praćen slojem sažimanja.

Konvolucija kao bitne detalje posmatra one koji su mnogo puta uhvaćeni u kernelu. Problem može da se desi kada kernel ne zahvata ivice dosta puta, te može mnogo da smanji određenu sliku, a samim tim i da se reši ivičnih detalja. Ako do te pojave dođe, koristi se tehnika koja se zove sužavanje. 

Sažimanje označava dodavanje piksela na ivice. Samim tim, kada konvolucija radi svoj posao, ona će svojim kernelom mnogo puta pokriti tu površinu. 

ReLU (rectified linear activation function / rectified linear unit) je funkcija koja negativnim vrednostima daje nulu, a pozitivne ostavlja kakve jesu. Time dobijamo nelinearan model.

![Funkcija](static\images\ReLU.svg)

Kroz neuronsku mrežu se propušta već napravljen spektrogram, kao i labele tih spektrograma koje mreža treba da raspozna.

Za treniranje mreže koriste se dve metode simultano (propagacija unapred i unazad), kao i jedna funkcija (kriterijumska funkcija)

Kritetijumska funkcija (eng. cost funkcija) jeste funkcionalna veza željenog outputa i dobijenog outputa u funkciji. Takođe, kriterijumska funkcija je usrednjena vrednost svih funkcija greške.

> Najkorišćenija loss funkcija je Cross Entropy Loss. Potrebno nam je da minimizujemo grešku unakrsne entropije za što preciznije rezultate.

Reviewer 6: ```suggestion
Najkorišćenija *loss* funkcija je *Cross Entropy Loss*. Potrebno nam je da minimizujemo grešku unakrsne entropije za što preciznije rezultate.
```

> Formula po kojoj se računa Cross Entropy Loss je sledeća:

Reviewer 6: ```suggestion
Formula po kojoj se računa *Cross Entropy Loss* je sledeća:
```

$$L_{C E}(\hat{y}, y)=-[y \log \sigma(\mathbf{w} \cdot \mathbf{x}+b)+(1-y) \log (1-\sigma(\mathbf{w} \cdot \mathbf{x}+b))]$$

Propagacija unazad je metod smanjenja grešaka u CNN posmatranjem neophodnih promena vrednosti parametara mreže u svakom sloju kako bi se neuroni aktivirali na određen način.

Propagacija unazad prolazi kroz sve slojeve i menja parametre mreže u svakom koraku.

> Parametri mreže se menjaju u cilju računanja dovoljno dobrog gradient spusta za traženje lokalnog / maksimalnog minimuma ove funkcije. Dakle, teži se tome da gradijent kriterijumske funkcije bude što bliži nuli.

Reviewer 6: ```suggestion
Parametri mreže se menjaju u cilju računanja dovoljno dobrog gradijentnog spusta za traženje lokalnog / maksimalnog minimuma ove funkcije. Dakle, teži se tome da gradijent kriterijumske funkcije bude što bliži nuli.
```

> Parametri mreže se menjaju u cilju računanja dovoljno dobrog gradient spusta za traženje lokalnog / maksimalnog minimuma ove funkcije. Dakle, teži se tome da gradijent kriterijumske funkcije bude što bliži nuli.

Reviewer 6: Koja je funkcija / u ovoj rečenici? "Lokalnog ili maksimalnog minimuma ove funkcije"? Zameniti sa odgovarajućom reči.

U FSDD bazi podataka, 6 osoba je izgovorila svaku reč 50 puta, kako bi ukupno bilo 3000 snimaka, što čini vrlo balansiranu bazu podataka. U srpskoj bazi, 10 reči je rečeno od strane 27 ljudi, dok su dve osobe ponovile izgovaranje ovih 10 reči 13 i 10 puta. Njihovih snimaka je 130 i 100, pa otuda i 500 snimaka u bazi. Korišćeni su drugačiji izgovori i intonacije zbog raznovrsnosti. Baza je slabije pristrasna, što se odražava na same rezultate testiranja.

U bazi srpskih reči, u uređenoj trojci gluva - glava - plava očekuju se češće greške pri klasifikaciji. To se može očekivati jer su drugi i poslednja dva glasa isti. Takođe, kako su „P“ i „G“ oba praskavi suglasnici, to jest isti su po mestu tvorbe, veća je verovatnoća pojavljivanja greške.

Za konvolucionu neuronsku mrežu, potrebni su nam bili pokazatelji kako mreža uči tokom epoha treniranja. Baze su podeljene na trening, test i validacionu bazu, tako da je trening set sadržao 70% reči, a test i validacioni set po 15% reči u slučaju obe baze.

> Rezultati su prikazani u tabeli ispod.
> 
> | Metoda | Logistička regresija | SVM | XGBoost | Random Forest Classificator | CNN |
> | -------- | -------- | -------- | -------- | -------- | -------- |
> | Tačnost | 66.33% | 70.11% | 85.90% | 82.67% | 97.28% |

Reviewer 6: Ja bih stavio da su metode u redovima a da je tačnost kolona. Ako nemate drugih metrika, razmislite o sledećem formatu:
```suggestion
Tačnosti ovih metoda bile su sledeće:

- **Logistička regresija:** 66.33%
- **SVM:** 70.11%
- **XGBoost:** 85.90%
- **Random Forest Classificator:** 82.67%
- **CNN:** 97.28%
```

Za konvolucionu neuronsku mrežu, potrebni su nam bili pokazatelji kako mreža uči tokom epoha treniranja. Baze su podeljene na trening, test i validacionu bazu, tako da je trening set sadržao 70% reči, a test i validacioni set po 15% reči u slučaju obe baze.

Rezultati su prikazani u tabeli ispod.

| Metoda | Logistička regresija | SVM | XGBoost | Random Forest Classificator | CNN |
| -------- | -------- | -------- | -------- | -------- | -------- |
| Tačnost | 66.33% | 70.11% | 85.90% | 82.67% | 97.28% |

Metrika ovih rezultata bila je tačnost. Zbog balansiranosti baze, ovo predstavlja zaista reprezentativnu metriku.

Odvojeno možemo posmatrati rezultate metoda sa dubokim učenjem i one bez dubokog učenja. Iz tabele se može uočiti da je konvoluciona neuronska mreža ostvarila najveću tačnost kao metoda sa dubokim učenjem, a XGBoost daje najbolje rezultate među metodama koje ne koriste duboko učenje.

> Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Metode sa dubokim učenjem same vrše feature extraction proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. Cross entropy loss, to jest log loss odlično funkcioniše kao loss funkcija za prepoznavanje govora pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~100Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.

Reviewer 6: Ovi engleski termini bi se mogli prevesti na srpski.
```suggestion
Konvoluciona neuronska mreža je metoda koja je najviše razrađena u ovom projektu. Metode sa dubokim učenjem same vrše *feature extraction* proces, koji je neophodan kako bismo sa spektrograma mogli lepo da izvučemo informacije o zvuku. *Cross entropy loss*, to jest *log loss* odlično funkcioniše kao *loss* funkcija za prepoznavanje govora pošto ljudsko uho reaguje logaritamski. To znači da je naše uho daleko osetljivije na niske frekvencije, primećujući razliku od svega nekoliko herca pri frekvencijama od ~100Hz, dok je ta razlika potpuno neprimetna na frekvencijama od nekoliko kHz. Osetljivost je pri dnu približno linearna, dok sa porastom frekvencije postaje logaritamska.
```

Rezultate vizuelno možemo prikazati matricama konfuzije. 

![Rezultati](static/images/XGB.svg)

![Rezultati](static/images/SVM1.svg)

![Rezultati](static/images/RandomForest.svg)

![Rezultati](static/images/LogistickaRegresija.svg)

Prikazane su matrice konfuzije na FSDD bazi za XGBoost, SVM, Random Forest i logističku regresiju, tim redosledom.

Iz ovih matrica konfuzije može se primetiti kako, nezavisno od metode koja se koristi, cifre dva, tri i četiri uvek imaju najveću tačnost pronalaženja. 

Cifre 9 i 1 su često mešane pri klasifikaciji kod ova četiri modela. Njihov izgovor se može protumačiti kao sličan ("one" i "nine"), te su ova dva broja par sa najvećim sličnostima u karakteristikama.

U svim metodama, cifra 6 je najviše puta pogrešno klasifikovana. Najčešće je mešana sa ciframa 3 i 8, što ima manje fizičkog smisla od mešanja cifara 1 i 9 ("six","three","eight").

> XGBoost i Random Forest su se pokazale kao najbolje metoda koje ne koriste tehniku dubokog učenja.  Obe navedene metode su ansambl metode koje koriste stabla odlučivanja, te su generalno veoma otporne na preprilagođavanje. Gledajući u tabelu, XGBoost je imao veću preciznost od Random Forest klasifikatora. Ovo se može objasniti "obrezivanjem drveća" koje XGBoost radi, to jest Boosting kojim poboljšava klasifikaciju. XGBoost u svakoj iteraciji pokušava da kompenzuje rezultate dosadašnjeg modela, te je bolji u prilagođavanju trening podacima.

Reviewer 6: ```suggestion
XGBoost i Random Forest su se pokazale kao najbolje metoda koje ne koriste tehniku dubokog učenja.  Obe navedene metode su ansambl metode koje koriste stabla odlučivanja, te su generalno veoma otporne na preprilagođavanje. Gledajući u tabelu, XGBoost je imao veću preciznost od Random Forest klasifikatora. Ovo se može objasniti "obrezivanjem drveća" koje XGBoost radi, to jest *Boosting*, kojim poboljšava klasifikaciju. XGBoost u svakoj iteraciji pokušava da kompenzuje rezultate dosadašnjeg modela, te je bolji u prilagođavanju trening podacima.
```

### Apstrakt na engleskom
### Uvod
Projekat "Prepoznavanje govora" pomaže pri rešavanju popularne dileme u AI tehnologiji, a to je kako da se glas pretvori u kucani tekst. – Prepoznavanje govora je proces osposobljavanja nekog modela da prepozna i odreaguje na zvuk proizveden ljudskim govorom. Model uzima audio signal u formi talasa, izvlači iz njega podatke, obrađuje ih, prepoznaje i preduzima određene korake u zavisnosti od rezultata.

> Motivacija projekta bila je u tome da se ne samo primene mnoge metode korišćene za speech recognition, već da se i uporede njihova praktičnost i tačnost. 

Reviewer: speech recognition -> prepoznavanje govora

> Primena prepoznavanja govora može se uočiti u mnogim svakodnevnim radnjama: audio pretraga na internetu, audio pretraga na uređajima za slepe ljude, voice dialing, ...

Reviewer: voice dialing -> pozivanje glasom / glasovno pozivanje

> Ovaj projekat se bavi prepoznavanjem konkretnih reči i njihova klasifikacija – ograničen je na prepoznavanje i klasifikaciju svega deset reči. Ceo projekat rađen je u Python programskom jeziku.

Reviewer: njihovom klasifikacijom * 
Nemojte da kazete projekat je ogranicen. Recite ovako:
Formulacija problema koji se resava u ovom projektu se moze definisati na sledeci nacin: Vrsi se klasifikacija reci na jednu od 10 reci iz predodredjenog skupa reci.

$c(x, y, f(x))= \begin{cases}0, & \text { if } y * f(x) \geq 1 \\ 1-y * f(x), & \text { else }\end{cases}$

> Ako su dobijeni i željeni rezultat istog znaka, vrednost kriterijumske funkcije je jednaka nuli, dok u suprotnom računamo gubitak. Na to moramo dodati i parametar za regularizaciju koji služi da izjednači uticaj maksimizacije granice i minimizacije gubitka.

Reviewer: Ne znam sta ste hteli da kazete u ovoj prvoj recenici, ali zvuci kao nesto sto je tacno samo a binarnu klasifikaciju. Obrisao bih tu recenicu
